### ISSUE TITLE
Matching accepts iterators with completely different values

### ISSUE THREAD

**Issue by peterflynn** [sinonjs/sinon#2400]
**Describe the bug**
If a function is called with an iterator, matching expected vs. actual arguments will accept any iterator of the same type – even when the values yielded by the two iterators are completely different. For example, the iterators from two different Sets with unequal contents are considered a match. This leads to tests that _think_ they are checking argument values, yet still pass even when the spy/stub is called with incorrect arguments.

Secondary issue, maybe separable: matching _won't_ accept an iterator with a different backing impl even when it _does_ yield the same values as the actual iterator. This can be challenging because tests may need to assume more about implementation details in other layers of code, or may not be able to instantiate the correct impl type at all. This is a bit less problematic since it's obvious you have a problem (the test you just wrote is failing) and I assume it can be worked around using a custom `sinon.match`-er?

**To Reproduce**

(runnable copy at https://codepen.io/peterflynn/pen/wvedJoK?editors=0012)

```js
const stub = sinon.stub();

// Call stub with iterator that yields ["one", "two"]
const actualSet = new Set(["one", "two"]);
stub(actualSet.keys());


// Passes even though this iterator yields different values!
const wrongSet = new Set(["four"]);
sinon.assert.calledWith(stub, wrongSet.keys());

// Fails even though this iterator yields the SAME values!
const correctButArray = ["one", "two"];
sinon.assert.calledWith(stub, correctButArray.values());


// Or put another way:
sinon.assert.match(actualSet.keys(),      wrongSet.keys());      // succeeds
sinon.assert.match([...actualSet.keys()], [...wrongSet.keys()]); // fails (as expected)

sinon.assert.match(actualSet.keys(),      correctButArray.values());      // fails
sinon.assert.match([...actualSet.keys()], [...correctButArray.values()]); // succeeds (as expected)
```

**Expected behavior**
* The `wrongSet` assert should either fail since the values yielded don't match, _or_ at minimum should fail with an error message saying iterator comparison isn't supported.

* Ideally the `correctButArray` assert should pass since the values yielded do match. My thinking is this feels a bit like comparing two objects with getters that yield the same value: the getter _impls_ may be different, but Sinon matching is ok with that as long as the getters yield the same value. Iterators _could_ be considered similar.

**Context (please complete the following information):**

- Library version: 9.2.4
- Environment:
- Example URL: https://codepen.io/peterflynn/pen/wvedJoK?editors=0012
- Other libraries you are using:

**Additional context**
n/a

**Comment by fatso83** [comment]
Great report, btw.

**Comment by rgroothuijsen** [comment]
One way to resolve this would be to have `samsam` treat a set iterator as if it were an array, but I figured I'd double-check if that is an acceptable solution.

**Comment by sdotson** [comment]
I've made an attempt here: https://github.com/sinonjs/samsam/pull/225 . Happy to address any and all feedback.

**Comment by sdotson** [comment]
Looks like my pr got merged. I think once there's a new `samsam` release we can update that dependency here in sinon and I'm thinking it'll probably just work at that point, but worth verifying that reality conforms to my expectations before closing this issue.

**Comment by fatso83** [comment]
@peterflynn I have verified that the iterator support is now working on your examples, but it only covers the first part. If the backing implementation is different, then the `deepEquals` implementation in `samsam` will return `false` before looking at the actual values. That means `actualSet.keys()` and `correctButArray.values()` are not deemed deeply equal.
![2022-02-03_11-22](https://user-images.githubusercontent.com/618076/152325274-5d7883d6-1755-481e-8df8-19d69c25fd7b.png)

This means we _have_ iterator support, but maybe you want to patch `samsam` with a tiny PR that treats iterators a bit differently in just this aspect.

Closing as support is in place.
